Local reference obliterate sided

titrindl opened the pull request:
Adds sidedness to `obliterateRange` endpoints. Instead of just number positions, start and end can be specified as `SequencePlace`s, allowing for a variety of different endpoint expansion behaviors.

The start place of the obliterated range can be:
1. Inclusive, where side === Side.Before (this is the default if a number position is provided)
2. Exclusive, where side === Side.After
The end place of the obliterated range can be:
1. Inclusive, where side === Side.After (this is the default if a number position is provided)
2. Exclusive, where side === Side.Before

When an exclusive endpoint is provided, concurrent inserts on that side of the obliterated range will be included in the obliteration.

anthony-murphy committed:
All test passing

anthony-murphy committed:
Merge branch 'main' of https://github.com/microsoft/FluidFramework into LocalReferenceObliterate

anthony-murphy committed:
Merge branch 'main' of https://github.com/microsoft/FluidFramework into LocalReferenceObliterate

anthony-murphy committed:
Merge branch 'main' of https://github.com/microsoft/FluidFramework into LocalReferenceObliterate

anthony-murphy committed:
Improve conditions in block insert to determine winning obliterate (#3)

* Improve conditions in block insert to determine winning obliterate

* improve ack and handle oldest vs newest obliterate

* fix reconnect

* PR feedback and build issues

* PR feedback and build issues

titrindl committed:
add sidedness

titrindl committed:
overlapping tests pass

titrindl committed:
add feature flag

titrindl committed:
Merge remote-tracking branch 'upstream/main' into LocalReferenceObliterate-sided

titrindl committed:
fix tests after merge

titrindl committed:
enable more tests

titrindl committed:
Merge remote-tracking branch 'upstream/main' into LocalReferenceObliterate-sided

anthony-murphy commented on packages/dds/merge-tree/src/client.ts:
```diff
@@ -252,15 +253,20 @@ export class Client extends TypedEventEmitter<IClientEvents> {
 	public obliterateRangeLocal(
-		start: number,
-		end: number,
+		start: SequencePlace,
```
we shouldn't use sequence place,  we should use `number | InteriorSequencePlace`, as we don't have and don't plan to add testing for `start | end`

anthony-murphy commented on packages/dds/merge-tree/src/mergeTree.ts:
```diff
@@ -90,9 +90,17 @@ import {
 import { SortedSegmentSet } from "./sortedSegmentSet.js";
 import { zamboniSegments } from "./zamboni.js";
 
+export function wasRemovedAfter(seg: ISegment, seq: number): boolean {
```
this was removed as it is no longer used, and i don't see a re-use

anthony-murphy commented on packages/dds/merge-tree/src/mergeTree.ts:
```diff
@@ -1896,8 +1922,222 @@ export class MergeTree {
+			// then it shouldn't be considered obliterated for other clients
+			// for refSeqs between the insertion's seq and the obliterate's seq.
+
+			if (segment.prevObliterateByInserter?.seq === UnassignedSequenceNumber) {
```
is this condition ever hit? I don't think the local client will ever run this method for previous obliterates

titrindl replied:
The condition is hit, I added a line to the comment explaining the circumstances where it is hit a bit more.

jzaffiro reviewed (commented):
The file isn't in this PR, but there are 2 tests in `obliterate.spec.ts` that will fail now, since they check that the obliterate range doesn't expand to include adjacently inserted segments. You can probably just delete them, since I think `rangeExpansion.spec.ts` takes care of inserts adjacent to where we actually do want to expand.

titrindl committed:
PR feedback

titrindl committed:
Update packages/dds/merge-tree/src/client.ts

remove empty comment line which happened to be close to a line touched by this PR

Co-authored-by: jzaffiro <110866475+jzaffiro@users.noreply.github.com>

titrindl commented:
> The file isn't in this PR, but there are 2 tests in `obliterate.spec.ts` that will fail now, since they check that the obliterate range doesn't expand to include adjacently inserted segments. You can probably just delete them, since I think `rangeExpansion.spec.ts` takes care of inserts adjacent to where we actually do want to expand.

@jzaffiro obliterate.spec.ts still passes, because I didn't turn on the sidedness feature flag there. Even if we do turn the sidedness feature flag on, those tests should continue to pass due to our decision to default to inclusive endpoints.

titrindl committed:
install deps after merge, build:fast passes

anthony-murphy commented on packages/dds/merge-tree/src/client.ts:
```diff
@@ -477,11 +483,21 @@ export class Client extends TypedEventEmitter<IClientEvents> {
 		const range = this.getValidOpRange(op, clientArgs);
-
+		if (this._mergeTree.options?.mergeTreeEnableSidedObliterate) {
+			start = { pos: op.pos1 ?? 0, side: op.before1 ? Side.Before : Side.After };
```
when will pos1 or pos2 be undefined? Is this a valid case, or should we make it an error?

anthony-murphy commented on packages/dds/merge-tree/src/client.ts:
```diff
@@ -587,8 +603,18 @@ export class Client extends TypedEventEmitter<IClientEvents> {
 			);
 		}
+		// eslint-disable-next-line import/no-deprecated
+		if (start !== undefined && (op as IMergeTreeObliterateMsg).before1 === false) {
```
we should do `as` casting when its possible to avoid, you can check the type of the op, which will narrow the type

anthony-murphy commented on packages/dds/merge-tree/src/mergeTree.ts:
```diff
@@ -1866,7 +1887,212 @@ export class MergeTree {
 	}
 
-	public obliterateRange(
+	private obliterateRangeSided(
```
this adds about 3kb to the bundle size? can we reduce the impact by sharing code?

anthony-murphy commented on packages/dds/merge-tree/src/mergeTree.ts:
```diff
@@ -1875,10 +2101,13 @@ export class MergeTree {
 		opArgs: IMergeTreeDeltaOpArgs,
 	): void {
-		errorIfOptionNotTrue(this.options, "mergeTreeEnableObliterate");
+		const startPlace = normalizePlace(start);
```
this isn't necessary, start and end are numbers, not need to normalize

anthony-murphy commented on packages/dds/merge-tree/src/opBuilder.ts:
```diff
@@ -88,8 +91,31 @@ export function createRemoveRangeOp(start: number, end: number): IMergeTreeRemov
+				: normalizePlace(end);
+		return {
+			pos1: startPlace.pos,
+			before1: startPlace.side === Side.Before,
```
i'm a little worried about modeling it this way, as a client who doesn't understand sideness will have no problem processing this op, it will just do it incorrectly by only processing pos1 and pos2 and ignoring before1 and before2, this will lead to eventually consistency errors which are quick hard to debug

anthony-murphy replied:
that, or allow pos1 and pos2 to be objects

anthony-murphy commented:
there is quite a lot in this PR. i think it would be good to break it apart where possible.

titrindl committed:
unify obliterate implementations
